src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.IAsyncEnumerable.cs

```diff
@@ -0,0 +1,303 @@
+                // produced no output
+                _target.ChangeBoundingCount(count: -1);
+            }
+            // else there's no reordering buffer, there are no output items, and we're not bounded,
+            // so there's nothing more to be done.
+        }
+
+        /// <summary>Stores the next item using the reordering buffer.</summary>
+        /// <param name="id">The ID of the item.</param>
+        /// <param name="item">The async enumerable.</param>
+        private async ValueTask StoreOutputItemsReorderedAsync(long id, IAsyncEnumerable<TOutput>? item)
+        {
+            Debug.Assert(_reorderingBuffer != null, "Expected a reordering buffer");
+            Debug.Assert(id != Common.INVALID_REORDERING_ID, "This ID should never have been handed out.");
+
+            // Grab info about the transform
+            TargetCore<TInput> target = _target;
+            bool isBounded = target.IsBounded;
+
+            // Handle invalid items (null enumerables) by delegating to the base
+            if (item == null)
+            {
+                _reorderingBuffer.AddItem(id, null!, false);
+                if (isBounded) target.ChangeBoundingCount(count: -1);
+                return;
+            }
+
+            // Determine whether this id is the next item, and if it is and if we have a trusted list,
+            // try to output it immediately on the fast path.  If it can be output, we're done.
+            // Otherwise, make forward progress based on whether we're next in line.
+            bool? isNextNullable = _reorderingBuffer.AddItemIfNextAndTrusted(id, null!, false);
+            if (!isNextNullable.HasValue) return; // data was successfully output
+            bool isNextItem = isNextNullable.Value;
+
+            // By this point, either we're not the next item, in which case we need to make a copy of the
+            // data and store it, or we are the next item and can store it immediately but we need to enumerate
+            // the items and store them individually because we don't want to enumerate while holding a lock.
+            List<TOutput>? itemCopy = null;
+            try
+            {
+                // If this is the next item, we can output it now.
+                if (isNextItem)
+                {
+                    await StoreOutputItemsNonReorderedWithIterationAsync(item).ConfigureAwait(false);
+                    // here itemCopy remains null, so that base.AddItem will finish our interactions with the reordering buffer
+                }
+                else
+                {
+                    // We're not the next item, and we're not trusted, so copy the data into a list.
+                    // We need to enumerate outside of the lock in the base class.
+                    int itemCount = 0;
+                    try
+                    {
+                        itemCopy = item.ToEnumerable().ToList(); // itemCopy will remain null in the case of exception
```

Comment:
We definitely do not want to do this.  It breaks the whole reason for having an IAsyncEnumerable.  This can just be:
```C#
itemCopy = new List<TOutput>();
await foreach (TOutput element in item.ConfigureAwait(false))
{
    itemCopy.Add(element);
}
```

Reply:
Great! So simple... I like it 😃